### DISCUSSION TITLE
Strugling with multiple workers setup

### DISCUSSION THREAD

**Discussion by pipozzz** [miguelgrinberg/Flask-SocketIO#1817] [Q&A]
Hello, 
I'd like to kindly ask about help with multiple workers setup.

I use gevent worker with gunicorn with -w 1 in from of this is nginx, and in front of this nginx I use aws ALB with stickysession enabled.
```
gunicorn --worker-class geventwebsocket.gunicorn.workers.GeventWebSocketWorker --max-requests 10000 --timeout 60 --keep-alive 60 --graceful-timeout 60 -w 1 -b 0.0.0.0:5000 --forwarded-allow-ips="*" run:app --enable-stdio-inheritance --access-logfile /proc/1/fd/1 --error-logfile /proc/1/fd/1 --daemon
```

flask-socketio I initialize like this:
```
socketio.init_app(app, debug=True, logger=True, cors_allowed_origins="*", engineio_logger=True,
message_queue=app.config['REDIS_URL'])
```

my nginx config
```
location / {
                proxy_pass http://127.0.0.1:5000;
                proxy_set_header Connection 'Upgrade';
                proxy_set_header Upgrade $http_upgrade;
                proxy_http_version 1.1;
                proxy_buffering off;
                proxy_cache off;
                proxy_redirect off;

                proxy_set_header   Host                 $host;
                proxy_set_header   X-Real-IP            $remote_addr;
                proxy_set_header   X-Forwarded-For      $proxy_add_x_forwarded_for;
                proxy_set_header   X-Forwarded-Proto    $http_x_forwarded_proto;
    }
```

When I have 1 worker it works fine, when I scale to 2 or so , then I'm getting 400 (Invalid session ID)

Do you have any idea where to look or what to adjust ?

Thanks for any help.

**Comment by miguelgrinberg** [comment]
The nginx configuration that you are showing here is for one worker, so I have no way to know what the problem is with multiple workers.

**Reply by pipozzz** [reply]
I have dockerized it so 1 docker (nginx + 1 gunicorn worker). 
Scaling means adding other dockers to ECS service. in AWS ALB I have enabled sticky sessions.

**Reply by miguelgrinberg** [reply]
Do you have a single Redis server/cluster for all your workers? Or are you also giving a separate redis to each one. The latter would not work, you need a single message queue used across the entire cluster of Socket.IO wokers.

**Reply by pipozzz** [reply]
I have single message queue for them

**Reply by miguelgrinberg** [reply]
I'm out of ideas then. Look in the log for any interesting messages that can indicate what's going on. You may need to look beyond error messages.

**Reply by pipozzz** [reply]
It seems that it's browser related issue, it's not working in Safari but works in Chrome, I'm debugging sticky sessions and it seems that Safari ignores cookie from AWS ALB (note: I'm connecting to different domain to API), but AWS ALB sets 2 cookies for that CORS situation but safari ignores it.

<img width="1002" alt="image" src="https://user-images.githubusercontent.com/1237882/164406281-b2bdba36-b071-47a7-a996-7ebb5e3d965a.png">

what do you think about disabling long-pooling so I could omit sticky sessions ?
Thanks for your help Miguel !

**Reply by miguelgrinberg** [reply]
The socket.io protocol gives the client the power to choose the initial transport, so you can change your clients to connect through websocket and avoid the sticky session requirement.

**Reply by pipozzz** [reply]
and does it have any disadvantage to omit long-pooling ?

**Reply by miguelgrinberg** [reply]
There are some, but are unrelated to Socket.IO. Browsers implement cross-origin protections for HTTP, but not for WebSocket (though this package does implement its own cross-origin protections in the server). The other one I can think of is that a WebSocket connection in the browser cannot carry custom headers, so auth via headers is not possible if you start with WS.

**Reply by pipozzz** [reply]
OK, thank for your help, it seems that I will do it with websocket transport only.
